### PR TITLE
Always revoke certificate on CRL

### PR DESCRIPTION
**Pull request by raoulstrackx**
## Description

Always revoke certificate on CRL
    
RFC5280 (https://tools.ietf.org/html/rfc5280) does not state that the `revocationDate` should be checked.
    
In addition, when no time source is available (i.e., when MBEDTLS_HAVE_TIME_DATE is not defined), `mbedtls_x509_time_is_past` always returns 0. This results in the CRL not being checked at all.

## Status
**READY**

## Requires Backporting
Yes 

- [x] 2.16 - #3562 
- [x] 2.7 - #3561 

## Todos

- [x] Tests
- [x] Changelog updated

Fix https://github.com/ARMmbed/mbedtls/issues/3340

### CODE REVIEW

**Comment by mpg** [comment]
Thanks for your contribution! I agree with your assessment: I looked at all instances of "revocationDate" in  RFC 5280 and it never says that it should be checked. Moreover, point (j) on [page 95](https://tools.ietf.org/html/rfc5280#page-95) says to marked as revoked as soon as issuer name and serial number match. It looks like this field is more informational (a bit like the optional "reason" field.) Also, from a security perspective, if the certificate ended up on a CRL, it shouldn't be trusted regardless of the revocation date, especially considering the local clock can't always be trusted.

The new behaviour should come with a test. Please generate a CRL with a revocation date in the far future (to make sure it's still in the future in a few years) for one of our existing certificate and add a test for verifying this certificate with that CRL using the function `x509_verify` in `tests/suites/test_suite_x509parse.data`. The generated CRL file should live in `tests/data_files` and its generation method documented in `tests/data_files/Makefile`. In order to obtain a revocation date in the future, you might need to use the `faketime` utility (see examples in that Makefile). Feel free to ask for help on testing if needed, we're aware that out test system is not always intuitive first time you touch it.

Since this changes the observable behaviour, it deserves an entry in our ChangeLog, I'd say under `Bugfix` as the current behaviour isn't in accordance to the standard. Can you write one? See [https://github.com/ARMmbed/mbedtls/blob/development/ChangeLog.d/00README.md](the ChangeLog README).

Finally, can you add a Signed-Off-By line to your commit message, ton indicate that you agree with our [licensing terms](https://github.com/ARMmbed/mbedtls#license)?

File: tests/suites/test_suite_x509parse.data
Comment (mpg) [inline tests/suites/test_suite_x509parse.data:914]: It took me some time to notice the difference between the two tests, could you perhaps append `HAVE_TIME_DATE`/`!HAVE_TIME_DATE` to the description in order to make it more apparent?

File: ChangeLog.d/fix-crl-revocationDate.txt
Comment (mpg) [inline ChangeLog.d/fix-crl-revocationDate.txt:2]: Typo: on longer -> no longer.

File: ChangeLog.d/fix-crl-revocationDate.txt
Comment (mpg) [inline ChangeLog.d/fix-crl-revocationDate.txt:5]: That's an interesting point. As you've noted when adding tests (thanks for that!), currently we have a `config.h` option `MBEDTLS_HAVE_TIME_DATE` whose documentation reads:

```
 * System has time.h, time(), and an implementation for
 * mbedtls_platform_gmtime_r() (see below).
 * The time needs to be correct (not necessarily very accurate, but at least
 * the date should be correct). This is used to verify the validity period of
 * X.509 certificates.
 *
 * Comment if your system does not have a correct clock.
```
Perhaps this documentation should be made more explicit:
```
 * \warning You should comment this out if the system's clock could be controlled by an attacker.
```

I'm mentioning this because my initial reaction was "but on systems where the local clock can't be trusted, people should just undefine `MBEDTLS_HAVE_TIME_DATE`" - then I checked if the documentation was actually making that clear, and perhaps not enough.

File: tests/suites/test_suite_x509parse.data
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_x509parse.data:914]: Generally, using numbers in test cases is deprecated unless it's for tests that do exactly the same thing with different (pseudo)random data.

I should get around to writing this guide on writing test case descriptions.

File: ChangeLog.d/fix-crl-revocationDate.txt
Comment (gilles-peskine-arm) [reply ChangeLog.d/fix-crl-revocationDate.txt:5]: “people should just undefine `MBEDTLS_HAVE_TIME_DATE`” That depends on your threat model. Undefining `MBEDTLS_HAVE_TIME_DATE` means that no certificate will ever be considered to have expired.

Comment (raoulstrackx) [reply ChangeLog.d/fix-crl-revocationDate.txt:5]: I agree description could be more clear. The same applies for `MBEDTLS_HAVE_TIME`: when the system's clock can't be trusted, it also cannot be trusted for measuring time differences.

Comment (raoulstrackx) [reply ChangeLog.d/fix-crl-revocationDate.txt:5]: Good point @gilles-peskine-arm but if you leave `MBEDTLS_HAVE_TIME_DATE` defined, things like SSL caches may never expire.

Comment (mpg) [reply ChangeLog.d/fix-crl-revocationDate.txt:5]: Ok, after thinking a bit more about this, I think I was wrong to suggest placing the ChangeLog entry under "Bugfix". As the PR description says, when `MBEDTLS_HAVE_TIME_DATE` is unset, we would simply never consider any certificate as revoked. I'm not sure how I overlooked this before, but that's enough to qualify as a security issue.

I'm now thinking the ChangeLog entry should look more like:

```
Security
   * When checking X.509 CRLs, a certificate was only considered as revoked if its revocationDate was in the past according to the local clock if available. In particular, on builds without MBEDTLS_HAVE_TIME_DATE, certificates were never considered as revoked. On builds with MBEDTLS_HAVE_TIME_DATE, an attacker able to control the local clock (for example, an untrusted OS attacking a secure enclave) could prevent revocation of certificates via CRLs. Fixed by no longer checking the revocationDate field, in accordance with RFC 5280.  Reported and fixed by Raoul Strackx & Jethro Beekman.
```

File: include/mbedtls/config.h
Comment (mpg) [inline include/mbedtls/config.h:149]: Ok, as Gilles points out, things are a bit more complex, and if we put this warning we should probably also put a warning that this option also prevents time-based expiration of certificates. Perhaps improving this documentation would be best done in another PR. (Or if done in this PR, it should definitely be a separate commit.)

File: tests/suites/test_suite_x509parse.data
Comment (mpg) [inline tests/suites/test_suite_x509parse.data:918]: Ok, sorry for failing to notice this initially, but I don't think we have any component in `tests/scripts/all.sh` that run tests in a build without `MBEDTLS_HAVE_TIME_DATE`, so currently this test if never run in our CI. We probably want to add a component to `all.sh` that builds with either the default configuration minus `MBEDTLS_HAVE_TIME_DATE`, or the full config minus that, and then runs `make test`.

File: ChangeLog.d/crl-revocationDate.txt
Comment (mpg) [inline ChangeLog.d/crl-revocationDate.txt:3]: `tests/scripts/all.sh` points out there is trailing whitespace on lines 3, 4 and 5 of this file. Please remove it.

**Comment by raoulstrackx** [comment]
@mpg could you send me the output of CI? Running the tests locally doesn't give me much insight. There are failed tests, but those fail on the development branch too.

**Comment by raoulstrackx** [comment]
One of the tests that fail is:

client:
```
../programs/ssl/ssl_client2 server_port=14928 server_addr=127.0.0.1 force_version=tls1_2 ca_file=none auth_mode=none crt_file=none key_file=none force_ciphersuite=TLS-ECDHE-ECDSA-WITH-AES-128-CBC-SHA256
                                                                                                                                                                                                                                             
  . Seeding the random number generator... ok                                                                                                                                                                                                
  . Loading the CA root certificate ... ok (0 skipped)                                                                                                                                                                                       
  . Loading the client cert. and key... ok (key type: invalid PK)                                                                                                                                                                            
  . Connecting to tcp/127.0.0.1/14928... ok                                                                                                                                                                                                  
  . Setting up the SSL/TLS structure... ok                                                                                                                                                                                                   
  . Performing the SSL/TLS handshake... failed                                                                                                                                                                                               
  ! mbedtls_ssl_handshake returned -0x7780                                                                                                                                                                                                   
                                                                                                                                                                                                                                             
Last error was: -0x7780 - SSL - A fatal alert message was received from our peer                                                                                                                                                             
                                                                                                                                                                                                                                             
EXIT: 1                                                                                                                                                                                                                                      
EXIT: 1
```

server:
```
gnutls-serv -p 14928 --http  --disable-client-cert --x509certfile data_files/server5.crt --x509keyfile data_files/server5.key --priority NORMAL:+AES-256-CCM-8:+AES-128-CCM-8:+ARCFOUR-128:+NULL:+MD5:+PSK:+DHE-PSK:+ECDHE-PSK:+RSA-PSK:-VERS-TLS-ALL:+VERS-TLS1.2                                                                                                                                                                                                                        
HTTP Server listening on IPv4 0.0.0.0 port 14928...done
HTTP Server listening on IPv6 :: port 14928...done                                                                                                                                                                                           
No certificates found!                                                                                                                                                                                                                       
Set static Diffie-Hellman parameters, consider --dhparams.                                                                                                                                                                                   
                                                                                                                                                                                                                                             
* Accepted connection from IPv4 127.0.0.1 port 51090 on Wed Jul  1 13:45:09 2020                                                                                                                                                             
- Description: (TLS1.2)-(ECDHE-ECDSA-SECP521R1)-(AES-128-CBC)-(SHA1)                                                                                                                                                                         
- Session ID: 3A:1B:0B:91:85:1B:B3:D2:D3:CA:5B:83:6B:AD:8C:B0:35:4A:21:01:72:E6:F5:B3:65:73:0D:F0:E0:C6:D3:21                                                                                                                                
- Given server name[1]: localhost                                                                                                                                                                                                            
- Ephemeral EC Diffie-Hellman parameters                                                                                                                                                                                                     
 - Using curve: SECP521R1                                                                                                                                                                                                                    
 - Curve size: 528 bits                                                                                                                                                                                                                      
- Version: TLS1.2
- Key Exchange: ECDHE-ECDSA
- Server Signature: ECDSA-SHA512
- Cipher: AES-128-CBC
- MAC: SHA1
- Compression: NULL
- Options: extended master secret, safe renegotiation, EtM,
- Channel binding 'tls-unique': 0c25dc0a33cd9291f2a8d0e0
No certificates found!

* Accepted connection from IPv4 127.0.0.1 port 51092 on Wed Jul  1 13:45:09 2020
- Description: (TLS1.2)-(ECDHE-ECDSA-SECP521R1)-(AES-256-CBC)-(SHA1)
- Session ID: 6F:38:69:0F:4D:1E:24:32:B2:1A:F6:E3:51:86:B9:1E:23:87:83:27:D3:A5:2A:26:29:6B:61:3B:FB:D6:60:39
- Given server name[1]: localhost
- Ephemeral EC Diffie-Hellman parameters
 - Using curve: SECP521R1
 - Curve size: 528 bits
- Version: TLS1.2
- Key Exchange: ECDHE-ECDSA
- Server Signature: ECDSA-SHA512
- Cipher: AES-256-CBC
- MAC: SHA1
- Compression: NULL
- Options: extended master secret, safe renegotiation, EtM,
- Channel binding 'tls-unique': f2fe159ed656e501fe4a5b85
Error in handshake
```

**Comment by gilles-peskine-arm** [comment]
You don't need to do anything else to pass CI.

* Travis passed.
* `pr-head` had a single test case failure which is in a DTLS test where an unexpected resend happened. This can happen if the OS drops a UDP packet, which is rare but possible. So that's as good as a pass.
* `pr-merge` passed with an older version of `development`, and didn't run with the latest version due to the "needs work" label. I've kicked a new run, should be over in an hour or so.
* Mbed OS jobs are failing due to an unrelated change on the Mbed OS side. We're ignoring Mbed OS CI results until that's fixed.

If you're getting failures with interoperability tests locally, it's probably because you're using different versions of OpenSSL and GnuTLS. Different tests need different versions because older versions lack some features or have bugs, and newer versions have removed some obsolete features. The docker image in `tests/docker` should work.

**Comment by raoulstrackx** [comment]
Oh that nice! Thanks for the fast reply!

**Comment by raoulstrackx** [comment]
Could this get a second review, please?

**Comment by ronald-cron-arm** [comment]
Hi, I am looking at this as the second reviewer and I have a question related to the first added test (with MBEDTLS_HAVE_TIME_DATE enabled). I was expecting that if I reverted the change in library/x509_crt.c (restoring the check of the revocation date) the test would fail. I tried but the test is passing (I've compiled the library and the tests with the default config and run the unit tests). What am I missing?

File: tests/suites/test_suite_x509parse.data
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_x509parse.data:916]: As @ronald-cron-arm noticed and I missed, this test doesn't do what it says. `crl-futureRevocationDate.pem` does not in fact have a revocation date in the future. Matching against [RFC 5280 §5.1](https://tools.ietf.org/html/rfc5280#section-5.1), `thisUpdate` and `nextUpdate` are in the future (2999-12-31 and 3010-01-01), but the revocation date for both serial numbers is 2011-02-12 12:44:07 which is in the past. Please fix the way `crl-futureRevocationDate.pem` is generated so that this becomes a proper non-regression test.

File: tests/data_files/Makefile
Comment (gilles-peskine-arm) [inline tests/data_files/Makefile:1058]: This looked right to me, but the result is not what we want: it sets the certificates' revocation dates to 2011-02-12 12:44:07 GMT. The current time (changed with faketime) and `-crldays` only influence the update times for the CRL, not the revocation date of the entries. The revocation date is coming from the start of the validity date of `test-ca.crt`. I guess it makes sense that the revocation of a certificate would be back-dated to the earliest time when it might have been generated. But it's not what we want here.

I don't know if there's a way to generate the test data that we want here with an OpenSSL command line. If there isn't, binary patching is acceptable as a last resort, as long as it's clearly explained in a comment.

File: ChangeLog.d/crl-revocationDate.txt
Comment (gilles-peskine-arm) [inline ChangeLog.d/crl-revocationDate.txt:10]: I just realized that the issue was reported independently in https://github.com/ARMmbed/mbedtls/issues/3340 (focusing on builds with `MBEDTLS_HAVE_TIME_DATE`).
```suggestion
     revocationDate field, in accordance with RFC 5280. Reported by
     yuemonangong in #3340. Reported independently and fixed by
     Raoul Strackx and Jethro Beekman in #3433.
```

File: tests/data_files/Makefile
Comment (jethrogb) [reply tests/data_files/Makefile:1058]: I think you can change the revocation date in `index.txt`.

**Comment by mpg** [comment]
The failures on Travis can be ignored, they're an instance of https://github.com/ARMmbed/mbedtls/issues/3520 and unrelated to this PR - the Mbed OS failures can also be ignored for similar reasons.

So this PR is all good, I think now's a good time to start preparing backports.

**Comment by raoulstrackx** [comment]
Thanks for your help!

Who would be backporting the patch?

**Comment by mpg** [comment]
Well, if you're willing to do it, that would be much appreciated. It would involve raising two new PRs, one targeting mbedtls-2.16, the other targeting mbedtls-2.7, with the same content as this one (usually `git cherry-pick` is helpful for that), and then linking to those in this PR's description.

Otherwise, we can handle it ourselves, but that may be a bit slower as that means involving more people on our side (1 for creating the PRs, 2 for reviewing them, while if you can create the PRs, the 2 reviewers who approved the original can usually review the backports pretty quickly).

Just let us know.

**Comment by raoulstrackx** [comment]
Sure I can do that.

**Comment by raoulstrackx** [comment]
Backported as #3561 and #3562

**Comment by gilles-peskine-arm** [comment]
Unfortunately, there's a conflict. It's a simple add/add conflict, but it prevents merging. Please either merge with the latest state of the `development` branch or rebase on top of `development`.

**Review by gilles-peskine-arm — APPROVED** [review]

**Comment by gilles-peskine-arm** [comment]
Thanks. Please resolve the similar conflict in the backports as well.

**Comment by raoulstrackx** [comment]
Sorry for the late response. Merge conflicts have been resolved

**Pull request merged**
